src/lightgbm/src/main/scala/LightGBMBooster.scala

```diff
@@ -35,81 +35,70 @@ class LightGBMBooster(val model: String) extends Serializable {
-      if (indexes.isDefined) lightgbmlib.delete_intArray(indexes.get._2)
-      if (indptrNative.isDefined) lightgbmlib.delete_intArray(indptrNative.get._2)
-      // Delete the output scored row
-      if (scoredDataOutPtr.isDefined) lightgbmlib.delete_doubleArray(scoredDataOutPtr.get)
-    }
+    val numCols = sparseVector.size
+
+    val datasetParams = "max_bin=255 is_pre_partition=True"
+    val dataInt32bitType = lightgbmlibConstants.C_API_DTYPE_INT32
+    val data64bitType = lightgbmlibConstants.C_API_DTYPE_FLOAT64
+
+    ensureScoredDataCreated
+
+    LightGBMUtils.validate(
+      lightgbmlib.LGBM_BoosterPredictForCSRSingle(
+        sparseVector.indices, sparseVector.values,
+        sparseVector.numNonzeros,
+        boosterPtr, dataInt32bitType, data64bitType, intToPtr(1 + 1), intToPtr(numCols),
+      kind, -1, datasetParams,
+      scoredDataLength_int64_tPtr, scoredDataOutPtr), "Booster Predict")
+
+    predToArray(classification, scoredDataOutPtr, kind)
   }
 
   protected def predictForMat(row: Array[Double], kind: Int, classification: Boolean): Array[Double] = {
-    var data: Option[(SWIGTYPE_p_void, SWIGTYPE_p_double)] = None
-    var scoredDataOutPtr: Option[SWIGTYPE_p_double] = None
-    try {
-      val numRows = 1
-      data = Some(LightGBMUtils.generateData(numRows, Array(row)))
-      val data64bitType = lightgbmlibConstants.C_API_DTYPE_FLOAT64
-
-      val numRowsIntPtr = lightgbmlib.new_intp()
-      lightgbmlib.intp_assign(numRowsIntPtr, numRows)
-      val numRows_int32_tPtr = lightgbmlib.int_to_int32_t_ptr(numRowsIntPtr)
-      val scoredDataLengthLongPtr = lightgbmlib.new_longp()
-      lightgbmlib.longp_assign(scoredDataLengthLongPtr, numRows)
-      val scoredDataLength_int64_tPtr = lightgbmlib.long_to_int64_t_ptr(scoredDataLengthLongPtr)
-
-      val numCols = row.length
-      val isRowMajor = 1
-      val numColsIntPtr = lightgbmlib.new_intp()
-      lightgbmlib.intp_assign(numColsIntPtr, numCols)
-      val numCols_int32_tPtr = lightgbmlib.int_to_int32_t_ptr(numColsIntPtr)
-      scoredDataOutPtr = Some(lightgbmlib.new_doubleArray(numRows * numClasses))
-      val datasetParams = "max_bin=255"
-      LightGBMUtils.validate(
-        lightgbmlib.LGBM_BoosterPredictForMat(
-          boosterPtr, data.get._1, data64bitType, numRows_int32_tPtr,
-          numCols_int32_tPtr, isRowMajor, kind,
-          -1, datasetParams, scoredDataLength_int64_tPtr, scoredDataOutPtr.get),
-        "Booster Predict")
-      predToArray(classification, scoredDataOutPtr.get, kind)
-    } finally {
-      // Delete the input row
-      if (data.isDefined) lightgbmlib.delete_doubleArray(data.get._2)
-      // Delete the output scored row
-      if (scoredDataOutPtr.isDefined) lightgbmlib.delete_doubleArray(scoredDataOutPtr.get)
-    }
+    val data64bitType = lightgbmlibConstants.C_API_DTYPE_FLOAT64
+
+    val numCols = row.length
+    val isRowMajor = 1
+
+    val datasetParams = "max_bin=255"
```

Comment:
Not related to this PR, but I just noticed the dataset param here is different from predictForCSR, which has is_pre_partition=True.
I think this is a bug.  Since the dataset is not distributed during prediction time, I think is_pre_partition shouldn't be set for CSR case during prediction, but I'm not sure why it isn't throwing some error right now.  Maybe it just ignores that param in this case, but we probably still shouldn't be setting it.